docs/_posts/2020-02-28-puppetwash.md

```diff
@@ -0,0 +1,381 @@
+wash docker/volumes/pupperware_puppetserver-config ❯ cp private_keys/puppet.test.pem ~/.puppetlabs/wash/pupperware/puppet.test.key.pem
+```
+
+Last, let's make sure there's something in PuppetDB to look at. In the Pupperware project we'll do a noop agent run to add facts and a report
+```
+$ docker-compose exec puppet puppet agent -t --noop
+Info: Using configured environment 'production'
+Info: Retrieving pluginfacts
+Info: Retrieving plugin
+Info: Retrieving locales
+Info: Applying configuration version '1582848507'
+Info: Creating state file /opt/puppetlabs/puppet/cache/state/state.yaml
+Notice: Applied catalog in 0.02 seconds
+```
+
+## Configure
+
+To configure Puppetwash we need to install and configure it. It's distributed as a Ruby Gem, so we can install it with
+```
+gem install puppetwash
+```
+
+Get the path to the Puppetwash script with
+```
+$ gem contents puppetwash --version 0.2.0
+/Users/me/.gem/ruby/2.6.0/gems/puppetwash-0.2.0/puppet.rb
+```
+> We specify `--version` to make sure we get the latest version we just installed.
+
+We then add that to Wash's config at `~/.puppetlabs/wash/wash.yaml`
+```
+external-plugins:
+- script: '/Users/me/.gem/ruby/2.6.0/gems/puppetwash-0.2.0/puppet.rb'
+```
+
+If you're using Puppet Enterprise you can use a user authentication (RBAC) token. You can generate a new one with
+```
+puppet access login -l 30d --print
+```
+
+Wash config would look like
+```
+external-plugins:
+- script: '/Users/me/.gem/ruby/2.6.0/gems/puppetwash-0.2.0/puppet.rb'
+puppet:
+  my_pe_instance:
+    puppetdb_url: https://puppet:8081
+    rbac_token: <my_rbac_token>
+    cacert: /path/to/cacert.pem
+```
+
+With open-source Puppet, we'll use cert-based authentication. Using the examples from above, `~/.puppetlabs/wash/wash.yaml` should look like
+```
+external-plugins:
+- script: '/Users/me/.gem/ruby/2.6.0/gems/puppetwash-0.2.0/puppet.rb'
+puppet:
+  my_pe_instance:
+    puppetdb_url: https://puppetdb.test:32770
+    cacert: /Users/michaelsmith/.puppetlabs/wash/pupperware/ca.pem
+    cert: /Users/michaelsmith/.puppetlabs/wash/pupperware/puppet.test.pem
+    key: /Users/michaelsmith/.puppetlabs/wash/pupperware/puppet.test.key.pem
+```
+
+## Explore
+
+Start (or restart) Wash to load the new Puppetwash config and you should now be able to explore Puppetwash
+```
+wash . > stree puppet
+puppet
+└── [instance]
+    └── nodes
+        └── [node]
+            ├── catalog.json
+            ├── facts
+            │   └── [fact]
+            └── reports
+                └── [report]
+```
+
+Puppetwash can connect to multiple instances of Puppet infrastructure (either a PE deployment or a combination of puppet masters and PuppetDB). It provides access to data in PuppetDB about nodes in that infrastructure. The example above configures one instance named `pupperware`
+```
+wash . > ls puppet
+pupperware/
+```
+
+Let's look at some PuppetDB data
+```
+wash . > cd puppet/pupperware/nodes
```

Comment:
I still think ls'ing the nodes before the `tree` would be useful. Then `tree` could show what's under a single node?

Reply:
Well, right now there is only a single node.

Reply:
Even then, still think ls followed by something like "`nodes` contains all your nodes" could help. But that is up to you.